### DISCUSSION TITLE
Permit Member Names to be the same as Class Names

### DISCUSSION THREAD

**Discussion by jhudsoncedaron** [dotnet/csharplang#3314] [Language Ideas]
Permit Member Names to be the same as Class Names

There's a lot of debate online about "Why do you want this?' Well, there are plenty of good reasons.

1) Object Mapper

    CREATE TABLE LocalizedText (
        LanguageId INTEGER NOT NULL,
        StringId INTEGER NOT NULL,
        LocalizedText NVARCHAR(max) NOT NULL
     );

The corresponding C# object really wants to be

      class LocalizedText {
           int LanguageId {get;set;}
           int StringId {get;set;}
           string LocalizedText {get;set;}
      }

While we originally debated syntax with methods, this really only comes up with methods on functor objects.

        public sealed class Next<T, TResult>
        {
            private readonly Func<T, Next, TResult> thisStep;
            private readonly Next next;

            public TResult Next(T @object) => thisStep(@object, next);

            public Next(Func<T, Next, TResult> processor, Next next)
            {
                thisStep = processor;
                this.next = next;
            }
        }

And there's the classic chestnut for autoconverting VB .NET:

       Public Class Main
              Public Shared app As MyApplication
              Public Shared frmMain as Form

              Public Shared Sub Main(ByVal args() As String)
                   Try
                       app = new MyApplication(args)
                       frmMain = app.CreateMainForm()
                       Application.Run(frmMain)
                   Catch ex As Exception
                       MessageBox.Show(ex.ToString())
                   End Try
              End Sub
       End Class

We used to have the following, but they don't exist anymore (when extract to library finally happened they ended up as two different methods in two different files on the same class):

      Public Class SignData
            Public Shared Function SignData(ByVal data as Byte(), ByVal key as Byte()) As Byte()
            End Function
      End Class

      Public Class VerifyData
            Public Shared Function VerifyData(ByVal data as Byte(), ByVal signature As Byte(), ByVal key as Byte()) As Boolean
            End Function
      End Class

**Comment by YairHalberstadt** [comment]
How would this impact name resolution?

For example:

```csharp

public class Foo
{
     public Bar Foo;
     public static int Bazz;
     public int M() => Foo.Bazz; // Is this the static field on the class `Foo` or the instance field in the field `Foo`?
}

public class Bar
{
    public int Bazz
}
```

**Comment by DaZombieKiller** [comment]
> 
> 
> How would this impact name resolution?
> 
> For example:
> 
> ```cs
> public class Foo
> {
>      public Bar Foo;
>      public static int Bazz;
>      public int M() => Foo.Bazz; // Is this the static field on the class `Foo` or the instance field in the field `Foo`?
> }
> 
> public class Bar
> {
>     public int Bazz
> }
> ```

I would imagine it would resolve to the static field, and the member could be accessed by using `this`. Regardless, I believe this illustrates exactly why things aren't done this way currently.

**Comment by jhudsoncedaron** [comment]
Name resolution would work like this case:

    public class Foo1
    {
        public Bar Foo;
    }

    public class Foo : Foo1
    {
        public static in Bazz;
        public int M() => Foo.Bazz;
    }

    public class Bar
    {
        public in Bazz
    }

All of my intended users are code generators that spit out C# code, so fully qualifying all the name references is no issue.
